flytepropeller/pkg/controller/nodes/attr_path_resolver.go

```diff
@@ -79,34 +88,131 @@ func resolveAttrPathInPbStruct(nodeID string, st *structpb.Struct, bindAttrPath
 	}
 
 	// After resolve, convert the interface to literal
-	literal, err := convertInterfaceToLiteral(nodeID, currVal)
+	literal, err := convertInterfaceToLiteral(nodeID, currVal, false)
+
+	return literal, err
+}
+
+// resolveAttrPathInJSON resolves the msgpack bytes (e.g. dataclass) with attribute path
+func resolveAttrPathInJSON(nodeID string, msgpackBytes []byte, bindAttrPath []*core.PromiseAttribute) (*core.Literal,
+	error) {
+
+	var currVal interface{}
+	var tmpVal interface{}
+	var exist bool
+	var jsonStr string
+
+	err := msgpack.Unmarshal(msgpackBytes, &jsonStr)
+	if err != nil {
+		return nil, err
+	}
+
+	// Golang has problem with unmarshalling integer as float64
+	// reference: https://stackoverflow.com/questions/22343083/json-unmarshaling-with-long-numbers-gives-floating-point-number
+
+	decoder := json.NewDecoder(strings.NewReader(jsonStr))
+	decoder.UseNumber()
+
+	err = decoder.Decode(&tmpVal)
+	if err != nil {
+		return nil, err
+	}
+	currVal = convertNumbers(tmpVal)
+
+	// Turn the current value to a map so it can be resolved more easily
+	for _, attr := range bindAttrPath {
+		switch resolvedVal := currVal.(type) {
+		// map
+		case map[string]interface{}:
+			tmpVal, exist = resolvedVal[attr.GetStringValue()]
+			if !exist {
+				return nil, errors.Errorf(errors.PromiseAttributeResolveError, nodeID, "key [%v] does not exist in literal %v", attr.GetStringValue(), currVal)
+			}
+			currVal = tmpVal
+		// list
+		case []interface{}:
+			if int(attr.GetIntValue()) >= len(resolvedVal) {
+				return nil, errors.Errorf(errors.PromiseAttributeResolveError, nodeID, "index [%v] is out of range of %v", attr.GetIntValue(), currVal)
+			}
+			currVal = resolvedVal[attr.GetIntValue()]
+		}
+	}
+
+	// After resolve, convert the interface to literal
+	literal, err := convertInterfaceToLiteral(nodeID, currVal, true)
 
 	return literal, err
 }
 
+// convertNumbers recursively converts json.Number to int64 or float64
+func convertNumbers(v interface{}) interface{} {
+	switch vv := v.(type) {
+	case map[string]interface{}:
+		for key, value := range vv {
+			vv[key] = convertNumbers(value)
+		}
+		return vv
+	case []interface{}:
+		for i, value := range vv {
+			vv[i] = convertNumbers(value)
+		}
+		return vv
+	case json.Number:
+		// Try to convert to int64 first
+		if intVal, err := vv.Int64(); err == nil {
+			return intVal
+		}
+		// If it fails, fall back to float64
+		if floatVal, err := vv.Float64(); err == nil {
+			return floatVal
+		}
+	}
+	return v
+}
+
 // convertInterfaceToLiteral converts the protobuf struct (e.g. dataclass) to literal
-func convertInterfaceToLiteral(nodeID string, obj interface{}) (*core.Literal, error) {
+func convertInterfaceToLiteral(nodeID string, obj interface{}, isJSON bool) (*core.Literal, error) {
```

Comment:
Could we add another function instead? `convertJsonToLiteral`
we can rename it to `convertStructToLiteral`